File: pipeline/v2/stage_0_validate.py
"""stage 0 — input validation and normalization. pure function, no api call."""

from prompts import VALID_MODES, VALID_VARIANTS
from voice_settings import VOICE_PRESET_MAP

LENGTH_LIMITS = {
    "intent": 100,
    "first_action": 80,
    "hero": 60,
    "phrase": 120,
}


def validate_and_normalize(inputs: dict) -> dict:
    """validate inputs and return a normalized copy."""
    out = dict(inputs)

    # required string fields
    for field in ("mode", "intent", "first_action", "hero", "phrase"):
        if field not in out or not isinstance(out[field], str):
            raise ValueError(f"input '{field}' is required and must be a string")
        out[field] = out[field].strip()
        if not out[field]:
            raise ValueError(f"input '{field}' cannot be empty after stripping whitespace")

    # length limits
    for field, limit in LENGTH_LIMITS.items():
        if len(out[field]) > limit:
            raise ValueError(
                f"input '{field}' is {len(out[field])} chars, exceeds limit of {limit}"
            )

    # mode normalization
    out["mode"] = out["mode"].lower()
    if out["mode"] not in VALID_MODES:
        raise ValueError(f"mode must be one of {VALID_MODES}, got: {out['mode']}")

    # variant
    out["variant"] = (out.get("variant") or "default").lower()
    if out["variant"] not in VALID_VARIANTS:
        raise ValueError(f"variant must be one of {VALID_VARIANTS}, got: {out['variant']}")

    # voice
    out["voice"] = (out.get("voice") or "the_closer").lower()
    if out["voice"] not in VOICE_PRESET_MAP:
        raise ValueError(
            f"voice must be one of {list(VOICE_PRESET_MAP.keys())}, got: {out['voice']}"
        )

    # hero: capitalize each word lightly (avoid mangling all-caps or middle-case inputs)
    if out["hero"] == out["hero"].lower():
        out["hero"] = " ".join(w.capitalize() for w in out["hero"].split())

    # target duration
    target = inputs.get("target_duration_seconds", 60)
    if not isinstance(target, int) or target < 30 or target > 120:
        target = 60
    out["target_duration_seconds"] = target

    return out

File: web/lib/oath-engine/stage-0-validate.ts
// stage 0: pure logic input validation. mirrors pipeline/v2/stage_0_validate.py

import 'server-only';
import promptsJson from './prompts.json';
import { VOICE_PRESET_MAP } from './voice-settings';
import type { PipelineInputs, ValidatedInputs, Mode, Variant, VoicePreset } from './types';

const LENGTH_LIMITS = {
  intent: 100,
  first_action: 80,
  hero: 60,
  phrase: 120,
} as const;

const VALID_MODES = promptsJson.valid_modes as readonly string[];
const VALID_VARIANTS = promptsJson.valid_variants as readonly string[];

export function validateAndNormalize(inputs: PipelineInputs): ValidatedInputs {
  for (const field of ['mode', 'intent', 'first_action', 'hero', 'phrase'] as const) {
    const v = inputs[field];
    if (typeof v !== 'string') {
      throw new Error(`input '${field}' is required and must be a string`);
    }
  }

  const intent = inputs.intent.trim();
  const first_action = inputs.first_action.trim();
  const hero = inputs.hero.trim();
  const phrase = inputs.phrase.trim();
  const mode = (inputs.mode as string).trim().toLowerCase();
  const variant = ((inputs.variant as string) || 'default').trim().toLowerCase();
  const voice = ((inputs.voice as string) || 'the_closer').trim().toLowerCase();

  if (!intent) throw new Error("input 'intent' cannot be empty");
  if (!first_action) throw new Error("input 'first_action' cannot be empty");
  if (!hero) throw new Error("input 'hero' cannot be empty");
  if (!phrase) throw new Error("input 'phrase' cannot be empty");

  if (intent.length > LENGTH_LIMITS.intent)
    throw new Error(`input 'intent' is ${intent.length} chars, exceeds limit of ${LENGTH_LIMITS.intent}`);
  if (first_action.length > LENGTH_LIMITS.first_action)
    throw new Error(
      `input 'first_action' is ${first_action.length} chars, exceeds limit of ${LENGTH_LIMITS.first_action}`
    );
  if (hero.length > LENGTH_LIMITS.hero)
    throw new Error(`input 'hero' is ${hero.length} chars, exceeds limit of ${LENGTH_LIMITS.hero}`);
  if (phrase.length > LENGTH_LIMITS.phrase)
    throw new Error(`input 'phrase' is ${phrase.length} chars, exceeds limit of ${LENGTH_LIMITS.phrase}`);

  if (!VALID_MODES.includes(mode))
    throw new Error(`mode must be one of ${VALID_MODES.join(', ')}, got: ${mode}`);
  if (!VALID_VARIANTS.includes(variant))
    throw new Error(`variant must be one of ${VALID_VARIANTS.join(', ')}, got: ${variant}`);
  if (!(voice in VOICE_PRESET_MAP))
    throw new Error(`voice must be one of ${Object.keys(VOICE_PRESET_MAP).join(', ')}, got: ${voice}`);

  const targetRaw = inputs.target_duration_seconds;
  const target = typeof targetRaw === 'number' && targetRaw >= 30 && targetRaw <= 120 ? targetRaw : 60;

  // lightweight hero capitalization if input is all-lowercase
  const heroNormalized =
    hero === hero.toLowerCase()
      ? hero.split(' ').map((w) => (w ? w[0].toUpperCase() + w.slice(1) : '')).join(' ')
      : hero;

  return {
    mode: mode as Mode,
    intent,
    first_action,
    hero: heroNormalized,
    phrase,
    variant: variant as Variant,
    voice: voice as VoicePreset,
    target_duration_seconds: target,
  };
}
